compiler/qsc_parse/src/lex/formatter.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::keyword::Keyword;
+
+use super::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind},
+    Delim, TokenKind,
+};
+use qsc_data_structures::span::Span;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug)]
+pub struct Edit {
+    #[allow(dead_code)] // TODO: nobody's using this yet except for tests
+    pub span: Span,
+    #[allow(dead_code)] // TODO: nobody's using this yet except for tests
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
+    "    ".repeat(level)
+}
+
+pub fn format(code: &str) -> Vec<Edit> {
+    let tokens = concrete::ConcreteTokenIterator::new(code);
+    let mut edits = vec![];
+
+    let mut indent_level = 0;
+
+    #[allow(unused_assignments)] // there's probably a better way of doing this, but this works
+    let mut one = None;
+    let mut two = None;
+    let mut three = None;
+
+    for token in tokens {
```

Comment:
It might be possible to do a larger refactor of the code in the future, including removing whitespace tokens in favor of a kind of whitespace trivia approach. Since this is under time pressure, let's table that possible refactoring for a future PR.

Reply:
I looked into doing this refactor and the logic to embed the whitespace spans into the tokens was much more complicated than I anticipated. The current structure is preferred for simplicity.